cody: new chat ui with stop generating button

abeatrix opened the pull request:
Close https://github.com/sourcegraph/sourcegraph/issues/53364 & https://github.com/sourcegraph/sourcegraph/issues/50058

- Update Chat UI based on design from Tim  ([figma](https://www.figma.com/proto/9112BsKsJc1BpO2j8XYwLL/%F0%9F%A4%96-Cody-VS-Code-%5BMain%5D?node-id=2803-99228&starting-point-node-id=2803%3A99228)) (Thanks @SuperAuguste  for the first draft!)
- Add stop generating button to abort a request in flight (Thanks @thenamankumar for adding it to web that I can use as references!)

#### Demo

https://github.com/sourcegraph/sourcegraph/assets/68532117/a37d22c8-dbee-4656-81e5-7cfedf5a950b

## Test plan

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

UI updated. Current tests are still passing:

```sh
cd client/cody
pnpm test:e2e
```

#### VS Code

<img width="1625" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/68532117/696b7355-f9af-4609-9232-56c8bd01142d">


https://github.com/sourcegraph/sourcegraph/assets/68532117/28715ecd-f68d-44f1-9ff7-3c5cfc66dd57

##### When Fixup is running 

https://github.com/sourcegraph/sourcegraph/assets/68532117/c37dba40-7a5a-4e6a-aa7b-5a009fdcc220

#### Web

<img width="2048" alt="Screenshot 2023-06-16 at 2 28 23 PM" src="https://github.com/sourcegraph/sourcegraph/assets/68532117/7d3d9be4-d280-4064-95ad-7845a2c78185">

<img width="2048" alt="Screenshot 2023-06-16 at 2 27 41 PM" src="https://github.com/sourcegraph/sourcegraph/assets/68532117/5d2e423a-2b5b-4d8b-a909-58d8e575a591">

https://github.com/sourcegraph/sourcegraph/assets/68532117/f1c87ce4-1a76-4785-8512-b49762eadde9

abeatrix commented on client/cody/webviews/Chat.module.css:
```diff
@@ -104,3 +107,20 @@ body[data-vscode-theme-kind='vscode-high-contrast'] .human-transcript-item {
+}
+
+.submit-button-disabled {
+    cursor: not-allowed;
```
show the button with `not-allowed` cursor to address feedback from users where they thought the button is clickable when it's not

toolmantim replied:
Ideally it would visually change too, but this works in a pinch.

I really like ChatGPT mobile’s interface, with the button transitioning to the stop button. I think that actually might be the better way for where we take this in the future.

abeatrix replied:
Yea i totally agree. I thought about updating that but i feel like i should do that in a different PR like you said in the future ;)

toolmantim commented:
I checked it out, and I really like where you've put the feedback and edit buttons. They look great!

Let's position the Stop Generating button fixed to the bottom, above the compose box:

https://github.com/sourcegraph/sourcegraph/assets/153/1d9452cc-97d9-4f76-bc1e-88e3a247ff16

I've transitioned the button in with a fade and small slide up, and a straight fade out to transition out. Let me know if you want me to help work on the transitions.

Here's the Figma (with clickable prototype):
https://www.figma.com/file/9112BsKsJc1BpO2j8XYwLL/%F0%9F%A4%96-Cody-VS-Code-%5BMain%5D?type=design&node-id=3532-142495&t=sGLSIMfj0YFYhiuN-4

(While I was here, I also designing the "..." animation. Maybe we can do that in a follow up? I can help with the CSS animations there too)

toolmantim commented:
@abeatrix are all the background shades okay to you? If there's any inconsistencies in the Figma to how we're using the VS Code theme color variables, let me know… because there's a good chance it was an accident/oversight, not by design.

Same for the "Stop Generating" button colours there… I used the small Figma VS Code button component, but the color and hover color etc should be the standard secondary one we're using.

abeatrix commented:
> https://www.figma.com/file/9112BsKsJc1BpO2j8XYwLL/%F0%9F%A4%96-Cody-VS-Code-%5BMain%5D?type=design&node-id=3532-142495&t=sGLSIMfj0YFYhiuN-4

> Let's position the Stop Generating button fixed to the bottom, above the compose box
I can update the styles for the button but i think we should update the position of the button in another PR so the web team can review, cause right now we are just creating the button for the editor that we can style freely and pass into the existing component that is shared between clients, which handles the position of the button. 

If we update the position of the button in the shared package, I think I will also need to update the button styles in the web client since it behaves differently than the editor one right now ( see video below, you needs to chase the button to stop it before the request is done :P). What do you think?

https://github.com/sourcegraph/sourcegraph/assets/68532117/c4ab924a-d363-48d9-95b4-69608f66acb8

valerybugakov commented on client/cody/webviews/Chat.tsx:
```diff
@@ -49,6 +49,14 @@ export const Chat: React.FunctionComponent<React.PropsWithChildren<ChatboxProps>
     suggestions,
     setSuggestions,
 }) => {
+    const [abortMessageInProgressInternal, setAbortMessageInProgress] = useState<() => void>(() => () => undefined)
```
What's the purpose of this state? It seems that the only value that it holds is the `noop` function.

abeatrix replied:
@valerybugakov good question! I saw web is using it plus it was one of the components required for the shared package to render the existing button component so I just copied the state that web is using 😅 

side note, I wrote the button component for vs code based on [your feedback in the original post](https://github.com/sourcegraph/sourcegraph/pull/53191#discussion_r1223693785), so thank YOU 😍

valerybugakov replied:
Gotcha! It looks like in the web implementation, we use this state [to reference custom abort functions](https://github.com/sourcegraph/sourcegraph/blob/acb8d71aab47762de71cc79afab20504e1b58467/client/cody-shared/src/chat/useClient.ts#L316). If we do not need it in VS Code, we can safely drop this state.

abeatrix replied:
You are right! Made the change and confirm the stop button is still working <3

valerybugakov reviewed (approved):
Nice!

toolmantim commented:
> If we update the position of the button in the shared package, I think I will also need to update the button styles in the web client since it behaves differently than the editor one right now ( see video below, you needs to chase the button to stop it before the request is done :P). What do you think?

I think we don't want people to be chasing buttons 😂

abeatrix commented:
> I think we don't want people to be chasing buttons 😂

haha I updated the button to get it fixed to the bottom, above the compose box, and added animation for transitioning in. i can't get it to fade out without adding a new state, so i just hold that off for now if that's ok.

> are all the background shades okay to you

I'm using the VS code webview tool kit for the editor components so they should just work...?  they look fine on my side at least 😅 

#### VS Code

<img width="1625" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/68532117/696b7355-f9af-4609-9232-56c8bd01142d">


https://github.com/sourcegraph/sourcegraph/assets/68532117/28715ecd-f68d-44f1-9ff7-3c5cfc66dd57

#### Web

<img width="1343" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/68532117/5062f305-645a-4b15-b7ea-bedc85b021d6">

https://github.com/sourcegraph/sourcegraph/assets/68532117/f1c87ce4-1a76-4785-8512-b49762eadde9

thenamankumar commented:
The gradient border on the left side of web looks a bit odd to me, given it will be of variable length. @almeidapaulooliveira can give approval from design side regarding this. 

Otherwise LGTM.

abeatrix commented:
@thenamankumar Thanks for doing all the work for web and for letting me copy your homework 🥹

dominiccooney reviewed (commented):
@abeatrix  I'm thinking that inline fixups, non-stop fixups will need a signal when the response is cancelled so that they don't sit around spinning forever or try to respin something that the user cancelled. (What happens when you hit cancel on them today?)

abeatrix commented:
> @abeatrix I'm thinking that inline fixups, non-stop fixups will need a signal when the response is cancelled so that they don't sit around spinning forever or try to respin something that the user cancelled. (What happens when you hit cancel on them today?)

@dominiccooney  I totally forgot about them 😅 I've updated so that `notifyTurnComplete` will be called when job is aborted. inline fixup will pick it up as error and display the error, while non-stop fixup will show no diff highlighted. I added a TODO for onError requests that I think we can work on in another PR :D

abeatrix commented:
@toolmantim @almeidapaulooliveira Can i get your review for the [Pending UI Review ](https://www.chromatic.com/review?appId=5f0f381c0e50750022dc6bf7&number=53332&type=linked)please?
![image](https://github.com/sourcegraph/sourcegraph/assets/68532117/eb2af333-77ec-4579-b9f3-b469e1a28785)

abeatrix commented:
Thanks @toolmantim ❤️
